File: sdks/go/httpclient.go
package modula

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

// httpClient is the internal HTTP transport shared by all resources on a
// [Client]. It handles base URL construction, Bearer token authentication,
// JSON encoding/decoding, and HTTP error mapping to [*ApiError].
//
// Timeout behavior is inherited from the underlying [http.Client] provided
// in [ClientConfig]. The default is 30 seconds. Context deadlines on
// individual requests take precedence over the client-level timeout.
type httpClient struct {
	baseURL    string
	apiKey     string
	httpClient *http.Client
}

// get performs a GET request to baseURL+path with optional query parameters.
// The response body is JSON-decoded into result. Pass nil for result to
// discard the body.
func (h *httpClient) get(ctx context.Context, path string, params url.Values, result any) error {
	fullURL := h.baseURL + path
	if len(params) > 0 {
		fullURL += "?" + params.Encode()
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fullURL, nil)
	if err != nil {
		return fmt.Errorf("modula: %w", err)
	}

	return h.do(req, result)
}

// post performs a POST request with a JSON-encoded body. The response is
// JSON-decoded into result.
func (h *httpClient) post(ctx context.Context, path string, body any, result any) error {
	return h.jsonRequest(ctx, http.MethodPost, path, body, result)
}

// put performs a PUT request with a JSON-encoded body. The response is
// JSON-decoded into result.
func (h *httpClient) put(ctx context.Context, path string, body any, result any) error {
	return h.jsonRequest(ctx, http.MethodPut, path, body, result)
}

// del performs a DELETE request with optional query parameters. The response
// body is discarded. Use delBody when the server returns data on deletion.
func (h *httpClient) del(ctx context.Context, path string, params url.Values) error {
	fullURL := h.baseURL + path
	if len(params) > 0 {
		fullURL += "?" + params.Encode()
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, fullURL, nil)
	if err != nil {
		return fmt.Errorf("modula: %w", err)
	}

	return h.do(req, nil)
}

// delBody performs a DELETE request and JSON-decodes the response body into
// result. Used for delete endpoints that return the deleted entity or a
// summary of cascaded deletions.
func (h *httpClient) delBody(ctx context.Context, path string, params url.Values, result any) error {
	fullURL := h.baseURL + path
	if len(params) > 0 {
		fullURL += "?" + params.Encode()
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, fullURL, nil)
	if err != nil {
		return fmt.Errorf("modula: %w", err)
	}

	return h.do(req, result)
}

// doRaw executes a pre-built request and returns the raw [http.Response].
// Authentication headers are set automatically. The caller is responsible
// for closing the response Body. This is used by specialized resources
// (e.g. MediaUpload) that need direct access to the response.
func (h *httpClient) doRaw(ctx context.Context, req *http.Request) (*http.Response, error) {
	req = req.WithContext(ctx)
	h.setAuth(req)

	resp, err := h.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("modula: %w", err)
	}

	return resp, nil
}

// jsonRequest builds and executes an HTTP request with a JSON-encoded body
// and Content-Type: application/json header. Used by post and put.
func (h *httpClient) jsonRequest(ctx context.Context, method string, path string, body any, result any) error {
	var buf bytes.Buffer
	if err := json.NewEncoder(&buf).Encode(body); err != nil {
		return fmt.Errorf("modula: encoding request body: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, method, h.baseURL+path, &buf)
	if err != nil {
		return fmt.Errorf("modula: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	return h.do(req, result)
}

// do is the central request executor. It sets the Authorization header,
// sends the request, checks the HTTP status code (returning [*ApiError]
// for any non-2xx response), and optionally JSON-decodes the response body
// into result. Pass nil for result to discard the body.
func (h *httpClient) do(req *http.Request, result any) error {
	h.setAuth(req)

	resp, err := h.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("modula: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return h.buildError(resp)
	}

	if result != nil {
		if err := json.NewDecoder(resp.Body).Decode(result); err != nil {
			return fmt.Errorf("modula: decoding response: %w", err)
		}
	}

	return nil
}

// setAuth adds the "Authorization: Bearer <apiKey>" header to the request.
// If no API key was configured, this is a no-op.
func (h *httpClient) setAuth(req *http.Request) {
	if h.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+h.apiKey)
	}
}

// buildError reads the response body and constructs an [*ApiError].
// It attempts to parse the body as JSON and extract a "message" or "error"
// field for a human-readable error description. If the body is not JSON or
// neither field is present, Message is left empty and the raw Body is still
// available for inspection.
func (h *httpClient) buildError(resp *http.Response) *ApiError {
	rawBody, readErr := io.ReadAll(resp.Body)

	apiErr := &ApiError{
		StatusCode: resp.StatusCode,
	}

	if readErr != nil {
		return apiErr
	}

	bodyStr := strings.TrimSpace(string(rawBody))
	apiErr.Body = bodyStr

	var parsed map[string]any
	if json.Unmarshal(rawBody, &parsed) == nil {
		if msg, ok := parsed["message"].(string); ok && msg != "" {
			apiErr.Message = msg
		} else if errMsg, ok := parsed["error"].(string); ok && errMsg != "" {
			apiErr.Message = errMsg
		}
	}

	return apiErr
}

File: sdks/swift/Sources/Modula/HTTPClient.swift
import Foundation

final class HTTPClient: Sendable {
    let baseURL: String
    let apiKey: String
    let session: URLSession

    init(baseURL: String, apiKey: String, session: URLSession) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.session = session
    }

    func get<T: Decodable>(path: String, queryItems: [URLQueryItem]? = nil) async throws -> T {
        let request = try buildRequest(method: "GET", path: path, queryItems: queryItems)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
        return try JSON.decoder.decode(T.self, from: data)
    }

    func post<T: Decodable, B: Encodable>(path: String, body: B) async throws -> T {
        let request = try buildJSONRequest(method: "POST", path: path, body: body)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
        return try JSON.decoder.decode(T.self, from: data)
    }

    func post<B: Encodable>(path: String, body: B) async throws {
        let request = try buildJSONRequest(method: "POST", path: path, body: body)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
    }

    func postNoBody(path: String) async throws {
        let request = try buildRequest(method: "POST", path: path)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
    }

    func put<T: Decodable, B: Encodable>(path: String, body: B) async throws -> T {
        let request = try buildJSONRequest(method: "PUT", path: path, body: body)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
        return try JSON.decoder.decode(T.self, from: data)
    }

    func patch<T: Decodable, B: Encodable>(path: String, body: B) async throws -> T {
        let request = try buildJSONRequest(method: "PATCH", path: path, body: body)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
        return try JSON.decoder.decode(T.self, from: data)
    }

    func delete(path: String, queryItems: [URLQueryItem]? = nil) async throws {
        let request = try buildRequest(method: "DELETE", path: path, queryItems: queryItems)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
    }

    func delete<T: Decodable>(path: String, queryItems: [URLQueryItem]? = nil) async throws -> T {
        let request = try buildRequest(method: "DELETE", path: path, queryItems: queryItems)
        let (data, response) = try await session.data(for: request)
        try checkStatus(response: response, data: data)
        return try JSON.decoder.decode(T.self, from: data)
    }

    func executeRaw(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        var req = request
        setAuth(&req)
        let data: Data
        let response: URLResponse
        if let body = req.httpBody {
            req.httpBody = nil
            (data, response) = try await session.upload(for: req, from: body)
        } else {
            (data, response) = try await session.data(for: req)
        }
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError(statusCode: 0, message: "Invalid response type")
        }
        return (data, httpResponse)
    }

    // MARK: - Private

    private func buildRequest(method: String, path: String, queryItems: [URLQueryItem]? = nil) throws -> URLRequest {
        var urlString = baseURL + path
        if let queryItems, !queryItems.isEmpty {
            var components = URLComponents()
            components.queryItems = queryItems
            if let query = components.percentEncodedQuery {
                urlString += "?" + query
            }
        }
        guard let url = URL(string: urlString) else {
            throw APIError(statusCode: 0, message: "Invalid URL: \(urlString)")
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        setAuth(&request)
        return request
    }

    private func buildJSONRequest<B: Encodable>(method: String, path: String, body: B) throws -> URLRequest {
        var request = try buildRequest(method: method, path: path)
        request.httpBody = try JSON.encoder.encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return request
    }

    private func setAuth(_ request: inout URLRequest) {
        if !apiKey.isEmpty {
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }
    }

    private func checkStatus(response: URLResponse, data: Data) throws {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError(statusCode: 0, message: "Invalid response type")
        }
        let statusCode = httpResponse.statusCode
        if statusCode < 200 || statusCode >= 300 {
            throw buildError(statusCode: statusCode, data: data)
        }
    }

    private func buildError(statusCode: Int, data: Data) -> APIError {
        let bodyStr = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        var message = ""
        if let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let msg = parsed["message"] as? String, !msg.isEmpty {
                message = msg
            } else if let errMsg = parsed["error"] as? String, !errMsg.isEmpty {
                message = errMsg
            }
        }

        return APIError(statusCode: statusCode, message: message, body: bodyStr)
    }
}
